driver/pace/driver/run.py

```diff
@@ -1,292 +1,379 @@
+        return datetime(2000, 1, 1)
+
+    def get_driver_state(
+        self,
+        quantity_factory: pace.util.QuantityFactory,
+        communicator: pace.util.CubedSphereCommunicator,
+    ) -> DriverState:
+        metric_terms = pace.util.grid.MetricTerms(
+            quantity_factory=quantity_factory, communicator=communicator
         )
-        quantity_factory = QuantityFactory.from_backend(sizer, backend=self._backend)
-        stencil_factory = StencilFactory(
-            config=stencil_config,
-            grid_indexing=grid_indexing,
+        state = pace.util.open_restart(
+            dirname=self.path,
+            communicator=communicator,
+            quantity_factory=quantity_factory,
         )
-        return quantity_factory, stencil_factory
+        raise NotImplementedError()
+
+
+@dataclasses.dataclass(frozen=True)
+class DiagnosticsConfig:
+    path: str
+    names: List[str] = dataclasses.field(default_factory=list)
+
 
-    def _compute_grid(self) -> Tuple["MetricTerms", "GridData"]:
-        metric_terms = MetricTerms(
-            quantity_factory=self._quantity_factory, communicator=self._comm
+class Diagnostics:
+    def __init__(
+        self,
+        config: DiagnosticsConfig,
+        partitioner: pace.util.CubedSpherePartitioner,
+        comm,
+    ):
+        self.config = config
+        store = zarr.storage.DirectoryStore(path=self.config.path)
+        self.monitor = pace.util.ZarrMonitor(
+            store=store, partitioner=partitioner, mpi_comm=comm
         )
-        grid_data = GridData.new_from_metric_terms(metric_terms)
-        return metric_terms, grid_data
-
-    def _initialize_dycore_state(
-        self, dycore_init_mode: Literal["serialized_data", "baroclinic"]
-    ) -> "DycoreState":
-        if dycore_init_mode == "serialized_data":
-            sys.path.append("/usr/local/serialbox/python/")
-            import serialbox
-
-            serializer = serialbox.Serializer(
-                serialbox.OpenModeKind.Read,
-                self._data_dir,
-                "Generator_rank" + str(self._comm.rank),
-            )
-            dycore_state = init_dycore_state_from_serialized_data(
-                serializer=serializer,
-                rank=self._comm.rank,
-                backend=self._backend,
-                namelist=self.dycore_config,
-                quantity_factory=self._quantity_factory,
-                stencil_factory=self._stencil_factory,
-            )
 
-        elif dycore_init_mode == "baroclinic":
-            # create an initial state from the Jablonowski & Williamson Baroclinic
-            # test case perturbation. JRMS2006
+    def store(self, time: datetime, state: DriverState):
+        zarr_state = {"time": time}
+        for name in self.config.names:
+            try:
+                quantity = getattr(state.dycore_state, name)
+            except AttributeError:
+                quantity = getattr(state.physics_state, name)
```

Comment:
Not right now - have you thought about how we should handle diagnostics variables inside physics?

Reply:
From the list of "would be nice to do" things from Lucas, it sounds like they'd like to be able to more easily add "deep diagnostics" in a somewhat performant way. The pattern I wrote here doesn't support that, but was easy to write for this example.

In the longer term, we probably want to design an API for registering (saying "hey I can compute this if you want it" from deep code), selecting (user configures which of those they actually want) and storing (when the code runs, the diagnostics need to be computed/saved) diagnostics.

We'd probably want to propagate a class through all of the init routines to do this, but there are other ways. One alternative is to have a context manager which sets the global diagnostic gatherer, like
```
with diagnostic_manager_1:
    driver_1.step()
with diagnostic_manager_2:
    driver_2.step()
```
But this would be a little less safe and a little more confusing.

Reply:
Thanks for the response. We should have this in our "would be nice to do" list.